std/node/_stream/end-of-stream.ts

```diff
@@ -0,0 +1,247 @@
+import type Writable from "./writable.ts";
+import {
+  ERR_INVALID_ARG_TYPE,
+  ERR_STREAM_PREMATURE_CLOSE,
+  NodeErrorAbstraction,
+} from "../_errors.ts";
+
+type Stream = Readable | Writable;
+
+function isRequest(stream: Stream) {
+  // deno-lint-ignore ban-ts-comment
+  //@ts-ignore
+  return stream.setHeader && typeof stream.abort === "function";
+}
+
+function isReadable(stream: Stream) {
+  // deno-lint-ignore ban-ts-comment
+  //@ts-ignore
+  return typeof stream.readable === "boolean" ||
+    // deno-lint-ignore ban-ts-comment
+    //@ts-ignore
+    typeof stream.readableEnded === "boolean" ||
+    // deno-lint-ignore ban-ts-comment
+    //@ts-ignore
+    !!stream._readableState;
+}
+
+function isWritable(stream: Stream) {
+  // deno-lint-ignore ban-ts-comment
+  //@ts-ignore
+  return typeof stream.writable === "boolean" ||
+    // deno-lint-ignore ban-ts-comment
+    //@ts-ignore
+    typeof stream.writableEnded === "boolean" ||
+    // deno-lint-ignore ban-ts-comment
+    //@ts-ignore
+    !!stream._writableState;
+}
+
+function isWritableFinished(stream: Writable) {
+  if (stream.writableFinished) return true;
+  const wState = stream._writableState;
+  if (!wState || wState.errored) return false;
+  return wState.finished || (wState.ended && wState.length === 0);
+}
+
+function nop() {}
+
+function isReadableEnded(stream: Readable) {
+  if (stream.readableEnded) return true;
+  const rState = stream._readableState;
+  if (!rState || rState.errored) return false;
+  return rState.endEmitted || (rState.ended && rState.length === 0);
+}
+
+interface FinishedOptions {
+  error?: boolean;
+  readable?: boolean;
+  writable?: boolean;
+}
+
+export default function eos(
+  stream: Stream,
+  options: FinishedOptions | null,
+  callback: (err?: NodeErrorAbstraction | null) => void,
+): () => void;
+export default function eos(
+  stream: Stream,
+  callback: (err?: NodeErrorAbstraction | null) => void,
+): () => void;
+
+export default function eos(
+  stream: Stream,
+  x: FinishedOptions | ((err?: NodeErrorAbstraction | null) => void) | null,
+  y?: (err?: NodeErrorAbstraction | null) => void,
+) {
+  let opts: FinishedOptions;
+  let callback: (err?: NodeErrorAbstraction | null) => void;
+
+  if (!y) {
+    if (typeof x !== "function") {
+      throw new ERR_INVALID_ARG_TYPE("callback", "function", x);
+    }
+    opts = {};
+    callback = x;
+  } else {
+    if (!x || Array.isArray(x) || typeof x !== "object") {
+      throw new ERR_INVALID_ARG_TYPE("opts", "object", x);
+    }
+    opts = x;
+
+    if (typeof y !== "function") {
+      throw new ERR_INVALID_ARG_TYPE("callback", "function", y);
+    }
+    callback = y;
+  }
+
+  callback = once(callback);
+
+  const readable = opts.readable ||
+    (opts.readable !== false && isReadable(stream));
+  const writable = opts.writable ||
+    (opts.writable !== false && isWritable(stream));
+
+  // deno-lint-ignore ban-ts-comment
+  //@ts-ignore
+  const wState = stream._writableState;
```

Comment:
I would rather see type assertions than ignores, even if the assertion is to `any`.

Reply:
Will work something out